src/gt4py/backend/base.py

```diff
@@ -386,64 +422,76 @@ class BaseModuleGenerator(abc.ABC):
 
     TEMPLATE_PATH = os.path.join(os.path.dirname(__file__), "templates", "stencil_module.py.in")
 
-    def __init__(self, builder: "StencilBuilder"):
-        assert isinstance(builder.backend, BaseBackend)
-        self.backend_class = builder.backend
-        self.builder = builder
-        self.options: gt_definitions.BuildOptions
-        self.stencil_id: gt_definitions.StencilID
-        self.definition_ir: gt_ir.StencilDefinition
+    builder: "StencilBuilder"
+    args_data: Dict[str, Any]
+    template: jinja2.Template
+
+    def __init__(self):
```

Comment:
Not initializing instance attributes in __init__ is problematic, because (even though they are declared for type checking) at runtime trying to access them before initialization only gives a "X has no Attribute Y" message.

Rather I will replace these with properties whose underlying attributes can be None but which raise a clearer error message in that case.

Reply:
In the case of `args_data` it will be sufficient to initialize it as empty dict for now.